src/core/api/immutabletagrule.go

```diff
@@ -0,0 +1,166 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/rbac"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/pkg/immutabletag"
+	"github.com/goharbor/harbor/src/pkg/retention/policy/rule"
+)
+
+// ImmutableTagRuleAPI ...
+type ImmutableTagRuleAPI struct {
+	BaseController
+	manager   immutabletag.RuleManager
+	projectID int64
+	ID        int64
+}
+
+// Prepare validates the user and projectID
+func (itr *ImmutableTagRuleAPI) Prepare() {
+	itr.BaseController.Prepare()
+	if !itr.SecurityCtx.IsAuthenticated() {
+		itr.SendUnAuthorizedError(errors.New("Unauthorized"))
+		return
+	}
+
+	pid, err := itr.GetInt64FromPath(":pid")
+	if err != nil || pid <= 0 {
+		text := "invalid project ID: "
+		if err != nil {
+			text += err.Error()
+		} else {
+			text += fmt.Sprintf("%d", pid)
+		}
+		itr.SendBadRequestError(errors.New(text))
+		return
+	}
+	itr.projectID = pid
+
+	ruleID, err := itr.GetInt64FromPath(":id")
+	if err == nil || ruleID > 0 {
+		itr.ID = ruleID
+	}
+
+	itr.manager = immutabletag.NewDefaultRuleManager()
+
+	if itr.SecurityCtx.IsSysAdmin() {
+		return
+	}
+
+	if strings.EqualFold(itr.Ctx.Request.Method, "get") {
```

Comment:
The flow will continue even though the checking fails, you should check the permission in the corresponding methods